aproxy.go

```diff
@@ -0,0 +1,404 @@
+				continue
+			}
+			if len(finalServerName) != 0 {
+				return "", fmt.Errorf("multiple names of the same name_type are prohibited in server name extension")
+			}
+			finalServerName = string(serverName)
+			if strings.HasSuffix(finalServerName, ".") {
+				return "", fmt.Errorf("SNI name ends with a trailing dot")
+			}
+		}
+	}
+	return finalServerName, nil
+}
+
+func PrereadHttpHost(conn *PrereadConn) (_ string, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("failed to preread HTTP request: %w", err)
+		}
+	}()
+
+	conn.StartPreread()
+	defer conn.RestorePreread()
+	req, err := http.ReadRequest(bufio.NewReader(conn))
+	if err != nil {
+		return "", err
+	}
+	host := req.Host
+	if host == "" {
+		return "", errors.New("http request doesn't have host")
+	}
+	return host, nil
+}
+
+func DialProxy(proxy string) (net.Conn, error) {
+	proxyAddr, err := net.ResolveTCPAddr("tcp", proxy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve proxy address: %w", err)
+	}
+	conn, err := net.DialTCP("tcp", nil, proxyAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func DialProxyConnect(proxy string, dst string) (net.Conn, error) {
+	conn, err := DialProxy(proxy)
+	if err != nil {
+		return nil, err
+	}
+	request := http.Request{
+		Method: "CONNECT",
+		URL: &url.URL{
+			Host: dst,
+		},
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Header: map[string][]string{
+			"User-Agent": {fmt.Sprintf("aproxy/%s", version)},
+		},
+		Host: dst,
+	}
+	err = request.Write(conn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to send connect request to http proxy: %w", err)
+	}
+	response, err := http.ReadResponse(bufio.NewReaderSize(conn, 0), &request)
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("proxy return %d response for connect request", response.StatusCode)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("failed to receive http connect response from proxy: %w", err)
+	}
+	return conn, nil
+}
+
+func GetOriginalDst(conn *net.TCPConn) (*net.TCPAddr, error) {
+	file, err := conn.File()
+	defer func(file *os.File) {
+		err := file.Close()
+		if err != nil {
+			slog.Error("failed to close the duplicated TCP socket file descriptor")
+		}
+	}(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert connection to file: %w", err)
+	}
+	var sockaddr [16]byte
+	size := 16
+	_, _, e := syscall.Syscall6(
+		syscall.SYS_GETSOCKOPT,
+		file.Fd(),
+		syscall.SOL_IP,
+		80, // SO_ORIGINAL_DST
+		uintptr(unsafe.Pointer(&sockaddr)),
+		uintptr(unsafe.Pointer(&size)),
+		0,
+	)
+	if e != 0 {
+		return nil, fmt.Errorf("getsockopt SO_ORIGINAL_DST failed: errno %d", e)
+	}
+	return &net.TCPAddr{
+		IP:   sockaddr[4:8],
+		Port: int(binary.BigEndian.Uint16(sockaddr[2:4])),
+	}, nil
+}
+
+func RelayTcp(conn io.ReadWriter, proxyConn io.ReadWriteCloser, logger *slog.Logger) {
+	var closed atomic.Bool
+	go func() {
```

Comment:
The lifetime of these goroutines is not determined. Is that an issue? Maybe it's not in a little utility like this, but normally you'd want something to ensure the caller waited for the goroutine to stop and exit. Like a `context.Context` or a done channel or something. Let me know if you want more info on this.

Reply:
The lifetime of this goroutine is tied to the lifetime of the incoming network connection. There are exactly two coroutines: one for reading and one for writing. The parent goroutine turns into the reading goroutine, and they return when the TCP connection is closed. I am actually not sure how I should control the lifetime of goroutines like this.

Reply:
That's fine -- it's a big topic, and what you have is probably fine for a simple CLI tool like this.

Reply:
I've come to think about this a bit, and controlling this goroutine seems difficult. First, it uses `io.Copy` to copy data, which doesn't accept a context. Second, the parent/caller goroutine also becomes the worker to read the connection, so there's no one waiting.
I wonder what is the proper or recommended way to handle this kind of situation in Go? Should the parent goroutine start two other goroutines to read/write and wait for them. Should we write our own implementation of `io.Copy` to accept context?

Reply:
If we have our own implementation of `io.Copy` we can handle the connection close gracefully instead of suppressing the error like now.